homeassistant/components/decora_wifi/light.py

```diff
@@ -2,90 +2,43 @@
     SUPPORT_TRANSITION,
     LightEntity,
 )
-from homeassistant.const import CONF_PASSWORD, CONF_USERNAME, EVENT_HOMEASSISTANT_STOP
-import homeassistant.helpers.config_validation as cv
+from homeassistant.const import CONF_USERNAME
+
+from .common import DecoraWifiEntity, DecoraWifiPlatform, decorawifisessions
 
 _LOGGER = logging.getLogger(__name__)
 
-# Validation of the user's configuration
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
-    {vol.Required(CONF_USERNAME): cv.string, vol.Required(CONF_PASSWORD): cv.string}
-)
 
-NOTIFICATION_ID = "leviton_notification"
-NOTIFICATION_TITLE = "myLeviton Decora Setup"
-
-
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up the Decora WiFi platform."""
-
-    email = config[CONF_USERNAME]
-    password = config[CONF_PASSWORD]
-    session = DecoraWiFiSession()
-
-    try:
-        success = session.login(email, password)
-
-        # If login failed, notify user.
-        if success is None:
-            msg = "Failed to log into myLeviton Services. Check credentials."
-            _LOGGER.error(msg)
-            hass.components.persistent_notification.create(
-                msg, title=NOTIFICATION_TITLE, notification_id=NOTIFICATION_ID
-            )
-            return False
-
-        # Gather all the available devices...
-        perms = session.user.get_residential_permissions()
-        all_switches = []
-        for permission in perms:
-            if permission.residentialAccountId is not None:
-                acct = ResidentialAccount(session, permission.residentialAccountId)
-                for residence in acct.get_residences():
-                    for switch in residence.get_iot_switches():
-                        all_switches.append(switch)
-            elif permission.residenceId is not None:
-                residence = Residence(session, permission.residenceId)
-                for switch in residence.get_iot_switches():
-                    all_switches.append(switch)
-
-        add_entities(DecoraWifiLight(sw) for sw in all_switches)
-    except ValueError:
-        _LOGGER.error("Failed to communicate with myLeviton Service")
-
-    # Listen for the stop event and log out.
-    def logout(event):
-        """Log out..."""
-        try:
-            if session is not None:
-                Person.logout(session)
-        except ValueError:
-            _LOGGER.error("Failed to log out of myLeviton Service")
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up Decora Wifi lights based on a config entry."""
+    # Retrieve the platform session from the reference stored in data.
+    email = entry.data[CONF_USERNAME]
+    session: DecoraWifiPlatform = decorawifisessions[email]
+
+    if session:
+        lights = session.lights
 
-    hass.bus.listen(EVENT_HOMEASSISTANT_STOP, logout)
+    entities = []
+    if lights:
+        for light in lights:
+            entities.append(DecoraWifiLight(light))
```

Comment:
use a list comprehension here

Reply:
Done in next commit.

Reply:
Reverted this in testing because I discovered that the list comprehension syntax I used resulted in the list comprehension's generator object being added to the entities list instead of iterating through and adding each entity. Will re-address tonight.